# 01 Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02 Import Data

In [2]:
df_orders = pd.read_csv(r"C:\Users\SigmaEd\Documents\CareerFoundry\Instacart Basket Analysis\02 Data\Original Data\orders.csv")
df_prods = pd.read_csv(r"C:\Users\SigmaEd\Documents\CareerFoundry\Instacart Basket Analysis\02 Data\Original Data\products.csv")
df_deps = pd.read_csv(r"C:\Users\SigmaEd\Documents\CareerFoundry\Instacart Basket Analysis\02 Data\Original Data\departments.csv")

# 03 Explore df_orders

In [3]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [12]:
df_orders.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [8]:
df_orders['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [7]:
df_orders.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
df_orders.tail()

order_id  user_id eval_set  order_number  order_dow  \
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0

# 04 Manipulate and Transform df_orders

In [16]:
# remove eval_set
df_orders_2 = df_orders.drop(columns = ['eval_set'])
print(df_orders_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   order_dow               int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB
None


In [17]:
# rename order_dow
df_orders_2.rename(columns = {'order_dow' : 'order_day_or_week'}, inplace = True)
print(df_orders_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   order_day_or_week       int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB
None


In [69]:
# change data type
df_orders_2['order_id'] = df_orders_2['order_id'].astype('str')
df_orders_2['user_id'] = df_orders_2['user_id'].astype('int')
print(df_orders_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int32  
 2   order_number            int64  
 3   order_day_or_week       int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int32(1), int64(3), object(1)
memory usage: 143.6+ MB
None


# new dataframe "df_orders_2"

# 05 Explore df_prods

In [20]:
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [24]:
df_prods.tail(10)

product_id                                product_name  aisle_id  \
49683       49679                     Famous Chocolate Wafers        61   
49684       49680          All Natural Creamy Caesar Dressing        89   
49685       49681     Spaghetti with Meatballs and Sauce Meal        38   
49686       49682                          California Limeade        98   
49687       49683                              Cucumber Kirby        83   
49688       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49689       49685          En Croute Roast Hazelnut Cranberry        42   
49690       49686                            Artisan Baguette       112   
49691       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49692       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49683             19     6.0  
49684             13     4.9  
49685              1     6.9  
49686              7     4.3  
49687              4    13.2  
49688              5     5.3  
49689              1     3.1  
49690              3     7.8  
49691              8     4.7  
49692             11    13.5

In [25]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# 06 Manipulate and Transform df_prods

In [59]:
# change data type
df_prods['product_id'] = df_prods['product_id'].astype('str')
df_prods['aisle_id'] = df_prods['aisle_id'].astype('str')
df_prods['department_id'] = df_prods['department_id'].astype('int')
print(df_prods.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  object 
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  object 
 3   department_id  49693 non-null  int32  
 4   prices         49693 non-null  float64
dtypes: float64(1), int32(1), object(3)
memory usage: 1.7+ MB
None


# 07 Explore df_deps

In [31]:
df_deps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  1 non-null      object
 1   1              1 non-null      object
 2   2              1 non-null      object
 3   3              1 non-null      object
 4   4              1 non-null      object
 5   5              1 non-null      object
 6   6              1 non-null      object
 7   7              1 non-null      object
 8   8              1 non-null      object
 9   9              1 non-null      object
 10  10             1 non-null      object
 11  11             1 non-null      object
 12  12             1 non-null      object
 13  13             1 non-null      object
 14  14             1 non-null      object
 15  15             1 non-null      object
 16  16             1 non-null      object
 17  17             1 non-null      object
 18  18             1 non-null      obj

In [32]:
df_deps.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# 08 Manipulate and Transform df_deps

In [44]:
# transpose

df_deps_t = df_deps.T
df_deps_t.reset_index()
new_header = df_deps_t.iloc[0]
df_deps_t_new = df_deps_t[1:]
df_deps_t_new.columns = new_header
print(df_deps_t_new)

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing


# New dataframe "df_deps_t_new"

# Question 4 - what the busiest hour is for placing orders

In [46]:
df_orders_2['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

# Question 4 Answer = 10 AM

# Question 5 - Meaning behind a value of 4 in the "department_id" column within the df_prods

In [52]:
# departments data dictionary
deps_data_dict = df_deps_t_new.to_dict('index')
deps_data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

# Question 5 Answer = Produce

# Question 6 - subset containing breakfast item sales

In [61]:
df_breakfast = df_prods[df_prods['department_id']==14]
df_breakfast

product_id                                       product_name aisle_id  \
27            28                                  Wheat Chex Cereal      121   
33            34                                                NaN      121   
67            68                            Pancake Mix, Buttermilk      130   
89            90                                       Smorz Cereal      121   
210          211   Gluten Free Organic Cereal Coconut Maple Vanilla      130   
...          ...                                                ...      ...   
49330      49326                            Cereal Variety Fun Pack      121   
49395      49391            Light and Fluffy Buttermilk Pancake Mix      130   
49547      49543                          Chocolate Cheerios Cereal      121   
49637      49633               Shake 'N Pour Buttermilk Pancake Mix      130   
49667      49663  Ultra Protein Power Crunch Peanut Butter N' Ho...       57   

       department_id  prices  
27                14    10.1  
33                14    12.2  
67                14    13.7  
89                14     3.9  
210               14     3.6  
...              ...     ...  
49330             14     9.1  
49395             14     2.0  
49547             14    10.8  
49637             14    14.2  
49667             14    10.2  

[1116 rows x 5 columns]

# Question 7 - subset containing alcohol, deli, beverages, and meat/seafood

In [63]:
df_dinner = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]
df_dinner

product_id                                    product_name aisle_id  \
2              3            Robust Golden Unsweetened Oolong Tea       94   
6              7                  Pure Coconut Water With Orange       98   
9             10  Sparkling Orange Juice & Prickly Pear Beverage      115   
10            11                               Peach Mango Juice       31   
16            17                               Rendered Duck Fat       35   
...          ...                                             ...      ...   
49676      49672                          Cafe Mocha K-Cup Packs       26   
49679      49675             Cinnamon Dolce Keurig Brewed K Cups       26   
49680      49676                          Ultra Red Energy Drink       64   
49686      49682                              California Limeade       98   
49688      49684       Vodka, Triple Distilled, Twist of Vanilla      124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Question 8 Answer = 7650

# Question 9 -  Extract all the information you can about "user_id" of “1.”

In [70]:
df_user1 = df_orders_2.loc[df_orders_2['user_id'].isin([1])]
df_user1

order_id  user_id  order_number  order_day_or_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   
5   3367565        1             6                  2                  7   
6    550135        1             7                  1                  9   
7   3108588        1             8                  1                 14   
8   2295261        1             9                  1                 16   
9   2550362        1            10                  4                  8   
10  1187899        1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# Question 10 - What basic stats can you provide based on the information you have

In [71]:
df_user1.describe()

user_id  order_number  order_day_or_week  order_hour_of_day  \
count     11.0     11.000000          11.000000          11.000000   
mean       1.0      6.000000           2.636364          10.090909   
std        0.0      3.316625           1.286291           3.477198   
min        1.0      1.000000           1.000000           7.000000   
25%        1.0      3.500000           1.500000           7.500000   
50%        1.0      6.000000           3.000000           8.000000   
75%        1.0      8.500000           4.000000          13.000000   
max        1.0     11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [72]:
df_user1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11 entries, 0 to 10
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   order_id                11 non-null     object 
 1   user_id                 11 non-null     int32  
 2   order_number            11 non-null     int64  
 3   order_day_or_week       11 non-null     int64  
 4   order_hour_of_day       11 non-null     int64  
 5   days_since_prior_order  10 non-null     float64
dtypes: float64(1), int32(1), int64(3), object(1)
memory usage: 572.0+ bytes


# Question 12 - Export df_orders_2

In [73]:
df_orders_2.to_csv(r"C:\Users\SigmaEd\Documents\CareerFoundry\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv")

# Question 13 - Export df_dep_t_new

In [74]:
df_deps_t_new.to_csv(r"C:\Users\SigmaEd\Documents\CareerFoundry\Instacart Basket Analysis\02 Data\Prepared Data\departments_wrangled.csv")